File: scripts/sanity_check.py
#!/usr/bin/env python3
"""判定口径的抽样自检：复现 panel.js 的核心链路，批量跑「多台风 × 多城市 × 多时刻」。

为什么要它：判定逻辑越来越多（相关半径、象限订正、窗口分离、分位标定），靠手点
页面测不过来，也测不了「不同时刻」。这个脚本用与前端**同源**的数据（温州台风网
路径 + Open-Meteo 逐时）复现 assess() 的关键步骤，可注入任意「现在时刻」，用于
回归与稳定性检查。

它不是前端的替身，只覆盖核心判据：相关性（含象限订正）、风雨窗口、已发生/未来
分离、档位、历史分位。

用法：
  python3 scripts/sanity_check.py 202613                 # 指定台风，当前时刻
  python3 scripts/sanity_check.py 202613 --sweep         # 叠加多时刻推演（稳定性）
  python3 scripts/sanity_check.py 202610 --at 2026-07-04T12 --cities 南宁市,柳州市
                                                          # 历史回放：注入当时时刻 +
                                                          # 用再分析实况，检验判据边界
"""
import json
import math
import sys
import urllib.request
from datetime import datetime, timezone, timedelta
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
BJT = timezone(timedelta(hours=8))
RAIN_ON, GUST_ON, GAP_H = 1.5, 62.0, 6

# 抽样城市：覆盖不同象限与距离（名字须与 regions.json 的城市键一致）
CITIES = ["杭州市", "上海市区", "宁波市", "南京市", "合肥市",
          "福州市", "武汉市", "广州市", "温州市", "青岛市"]


def get(url, timeout=45):
    req = urllib.request.Request(url, headers={"User-Agent": "typhoonandcicada-selftest"})
    return json.loads(urllib.request.urlopen(req, timeout=timeout).read().decode("utf-8"))


def hav(a, b, c, d):
    R, p = 6371.0, math.pi / 180
    x = math.sin((c - a) * p / 2) ** 2 + math.cos(a * p) * math.cos(c * p) * math.sin((d - b) * p / 2) ** 2
    return 2 * R * math.asin(math.sqrt(x))


def bearing(a, b, c, d):
    p = math.pi / 180
    y = math.sin((d - b) * p) * math.cos(c * p)
    x = math.cos(a * p) * math.sin(c * p) - math.sin(a * p) * math.cos(c * p) * math.cos((d - b) * p)
    return (math.degrees(math.atan2(y, x)) + 360) % 360


def est_gale(pw):
    pw = int(pw or 0)
    return (400 if pw >= 16 else 350 if pw >= 14 else 300 if pw >= 12 else
            230 if pw >= 10 else 160 if pw >= 8 else 110 if pw >= 6 else 70)


def ptime(s):
    s = s.replace("T", " ").split("+")[0].split(".")[0]
    return datetime.strptime(s, "%Y-%m-%d %H:%M:%S").replace(tzinfo=BJT).timestamp()


def load_city_coords():
    reg = json.loads((ROOT / "docs" / "data" / "regions.json").read_text(encoding="utf-8"))
    out = {}
    for prov, pv in reg.items():
        for city, cv in (pv.get("cities") or {}).items():
            out[city] = (cv["lat"], cv["lng"])
    return out


def load_storm(tfid):
    d = get(f"https://typhoon.slt.zj.gov.cn/Api/TyphoonInfo/{tfid}")
    pts = d.get("points", [])
    fcs = {}
    for i in range(len(pts) - 1, -1, -1):
        for fc in pts[i].get("forecast") or []:
            if fc.get("tm") and fc["tm"] not in fcs and fc.get("forecastpoints"):
                fcs[fc["tm"]] = fc["forecastpoints"]
        if fcs:
            break
    fc = fcs.get("中国") or (list(fcs.values())[0] if fcs else [])
    return d.get("name"), pts, fc


def load_wx_archive(lat, lng, at_ep):
    """历史回放用 ERA5 再分析实况（archive API）：取 [at−14d, at+7d]。
    注意这是**实况回放**，检验的是「判据在真实天气下会说什么」，
    不是「当时的预报准不准」（我们没有存历史预报）。"""
    s = datetime.fromtimestamp(at_ep - 14 * 86400, BJT).strftime("%Y-%m-%d")
    e = datetime.fromtimestamp(at_ep + 7 * 86400, BJT).strftime("%Y-%m-%d")
    d = get(f"https://archive-api.open-meteo.com/v1/archive?latitude={lat}&longitude={lng}"
            f"&start_date={s}&end_date={e}&hourly=precipitation,wind_gusts_10m"
            f"&daily=precipitation_sum&timezone=Asia%2FShanghai")
    h = d["hourly"]
    t = [datetime.strptime(x, "%Y-%m-%dT%H:%M").replace(tzinfo=BJT).timestamp() for x in h["time"]]
    dl = d.get("daily", {})
    days, sums = dl.get("time", []), dl.get("precipitation_sum", [])
    cutoff = datetime.fromtimestamp(at_ep, BJT).strftime("%Y-%m-%d")
    pa = 0.0
    past = [(dd, v) for dd, v in zip(days, sums) if dd < cutoff and v is not None]
    for i, (_dd, v) in enumerate(reversed(past)):
        pa += (0.85 ** i) * v
    return t, h["precipitation"], h["wind_gusts_10m"], min(1.0, pa / 100.0)


def load_wx(lat, lng):
    """逐时风雨 + 前期影响雨量 Pa 的土壤饱和度。
    一次取 14 天（前端是 hourly 7 天 + daily 14 天两次请求），这里合并成一次；
    窗口搜索同样限定在最近点 ±36/48h，故与前端结果一致。"""
    d = get(f"https://api.open-meteo.com/v1/forecast?latitude={lat}&longitude={lng}"
            f"&hourly=precipitation,wind_gusts_10m&daily=precipitation_sum"
            f"&past_days=14&forecast_days=7&timezone=Asia%2FShanghai")
    h = d["hourly"]
    t = [datetime.strptime(x, "%Y-%m-%dT%H:%M").replace(tzinfo=BJT).timestamp() for x in h["time"]]
    # Pa = Σ Kⁱ·P(i天前)，K=0.85；w = min(1, Pa/100)——与 panel.js 同口径
    dl = d.get("daily", {})
    days, sums = dl.get("time", []), dl.get("precipitation_sum", [])
    today = datetime.now(BJT).strftime("%Y-%m-%d")
    pa = 0.0
    past = [(dd, v) for dd, v in zip(days, sums) if dd < today and v is not None]
    for i, (_dd, v) in enumerate(reversed(past)):      # i=0 是昨天
        pa += (0.85 ** i) * v
    w = min(1.0, pa / 100.0)
    return t, h["precipitation"], h["wind_gusts_10m"], w


def assess(lat, lng, pts, fc, wx, now_ep, pct_tab):
    """复现 panel.js assess() 的核心判据。"""
    seq = [{"lat": float(p["lat"]), "lng": float(p["lng"]), "power": p.get("power"),
            "ep": ptime(p["time"])} for p in pts]
    seq += [{"lat": float(p["lat"]), "lng": float(p["lng"]), "power": p.get("power"),
             "ep": ptime(p["time"])} for p in fc]
    for p in seq:
        p["dist"] = hav(lat, lng, p["lat"], p["lng"])
    closest = min(seq, key=lambda p: p["dist"])

    wr = est_gale(closest["power"])
    # 象限订正的降雨相关半径
    prev = [p for p in seq if p["ep"] <= closest["ep"]]
    nxt = [p for p in seq if p["ep"] > closest["ep"]]
    a_, b_ = (prev[-1] if prev else seq[0]), (nxt[0] if nxt else seq[-1])
    dirf = None
    if a_ is not b_:
        mv = bearing(a_["lat"], a_["lng"], b_["lat"], b_["lng"])
        rel = (bearing(closest["lat"], closest["lng"], lat, lng) - mv + 360) % 360
        dirf = 0.65 + 0.35 * math.cos(math.radians(rel - 45))
    base = min(600, max(300, wr * 2.5))
    rr = base if dirf is None else max(150, base * dirf)
    in_range = [p for p in seq if p["dist"] < est_gale(p["power"])]
    relevant = bool(in_range) or closest["dist"] <= rr

    # 风雨窗口（含最近时刻的那段）
    t, pr, gu, soil_w = wx
    rain = past = future = 0
    win = None
    if relevant:
        anchor = closest["ep"]
        lo, hi = anchor - 36 * 3600, anchor + 48 * 3600
        segs, cur = [], None
        for i, tt in enumerate(t):
            if tt < lo or tt > hi:
                continue
            if (pr[i] or 0) >= RAIN_ON or (gu[i] or 0) >= GUST_ON:
                if cur and tt - t[cur[1]] > GAP_H * 3600:
                    segs.append(cur); cur = None
                cur = (i, i) if not cur else (cur[0], i)
        if cur:
            segs.append(cur)
        seg = next((s for s in segs if t[s[0]] <= anchor <= t[s[1]] + GAP_H * 3600), None)
        if not seg and segs:
            seg = min(segs, key=lambda s: min(abs(t[s[0]] - anchor), abs(t[s[1]] - anchor)))
        if seg:
            win = (t[seg[0]], t[seg[1]])
            for i in range(seg[0], seg[1] + 1):
                v = pr[i] or 0
                rain += v
                if t[i] <= now_ep:
                    past += v
                else:
                    future += v
    rain, past, future = round(rain), round(past), round(future)

    # 档位（与 panel.js 同口径）：雨臂随土壤湿度下调；风臂用**本地实际峰值阵风**
    pw = int(closest["power"] or 0)
    wet = 1 - 0.4 * soil_w
    level = 1
    if rain >= 60 * wet:
        level = 2
    if rain >= 150 * wet:
        level = 3
    if rain >= 250:                      # 档4 不随湿度下调
        level = 4
    pgust = None
    if win:
        i0 = next(i for i, tt in enumerate(t) if tt >= win[0])
        i1 = next(i for i in range(len(t) - 1, -1, -1) if t[i] <= win[1])
        pgust = max((gu[i] or 0) for i in range(i0, i1 + 1)) if i1 >= i0 else None
    if pgust is not None:
        if pgust >= 89:  level = max(level, 2)     # 阵风10级·黄
        if pgust >= 118: level = max(level, 3)     # 阵风12级·橙
        if pgust >= 150: level = max(level, 4)     # 阵风14级·红
    else:
        if closest["dist"] < wr and pw >= 8:   level = max(level, 2)
        if closest["dist"] < 200 and pw >= 10: level = max(level, 3)
    if closest["dist"] < 100 and pw >= 14:     level = max(level, 4)   # 安全网

    # 阶段
    phase = "approach"
    if win and now_ep >= win[0]:
        phase = "after" if now_ep > win[1] else "during"

    pct = None
    if pct_tab and rain > 0:
        rec = pct_tab.get("_city")
        if rec and len(rec["v"]) >= 10:
            pct = round(100 * sum(1 for v in rec["v"] if v <= rain) / len(rec["v"]))
    # 内涝清单是否会推给「没勾低洼/商铺」的普通城市居民（与 panel.js 同口径）
    flood = bool(relevant and rain >= 50 * wet and phase != "after")
    # 本地风雨持续时长，及「风雨持续时间长」提示是否触发（与 panel.js 同口径）
    dur_h = (win[1] - win[0]) / 3600 if win else None
    accum_heavy = rain >= 50 * wet
    long_rain = bool(relevant and dur_h is not None and dur_h >= 24 and accum_heavy)
    slow_threat = bool(closest["dist"] < wr and accum_heavy)   # 移速另算，此处只看几何+雨量
    return dict(dist=round(closest["dist"]), pw=pw, rr=round(rr), dirf=dirf,
                relevant=relevant, rain=rain, past=past, future=future, soil=round(soil_w, 2),
                level=level, phase=phase, pct=pct, win=win, flood=flood,
                durH=dur_h, longRain=long_rain, gust=round(pgust) if pgust else None)


def _opt(name, default=None):
    for i, a in enumerate(sys.argv):
        if a == name and i + 1 < len(sys.argv):
            return sys.argv[i + 1]
    return default


def main():
    args = [a for a in sys.argv[1:] if not a.startswith("--")]
    args = [a for a in args if a not in (_opt("--at"), _opt("--cities"))]
    sweep = "--sweep" in sys.argv
    tfid = args[0] if args else "202613"
    at = _opt("--at")            # 历史回放时刻，如 2026-07-04T12
    cities_opt = _opt("--cities")
    coords = load_city_coords()
    pct_all = json.loads((ROOT / "docs" / "data" / "rain-percentile.json").read_text(encoding="utf-8"))["d"]
    name, pts, fc = load_storm(tfid)
    if at:
        now = datetime.strptime(at, "%Y-%m-%dT%H").replace(tzinfo=BJT).timestamp()
    else:
        now = datetime.now(timezone.utc).timestamp()
    city_list = [c.strip() for c in cities_opt.split(",")] if cities_opt else CITIES
    tag = "历史回放" if at else "现在"
    print(f"=== {name} ({tfid}) · {tag} {datetime.fromtimestamp(now, BJT):%Y-%m-%d %H:%M} 北京时 ===\n")
    print(f"{'城市':<7}{'距离':>7}{'雨半径':>7}{'向':>5}{'相关':>5}{'过程':>6}{'已下':>6}{'土湿':>6}{'阵风':>8}{'档':>3}{'阶段':>7}{'持续':>7}{'长雨提示':>9}")
    print("-" * 72)
    rows = []
    for c in city_list:
        if c not in coords:
            print(f"  {c}: 不在区县库中，跳过")
            continue
        lat, lng = coords[c]
        try:
            wx = load_wx_archive(lat, lng, now) if at else load_wx(lat, lng)
        except Exception as e:
            print(f"{c:<7} 降水拉取失败: {e}")
            continue
        tab = {"_city": pct_all.get(c)}
        r = assess(lat, lng, pts, fc, wx, now, tab)
        rows.append((c, lat, lng, wx, r))
        d = f"{r['dirf']:.2f}" if r["dirf"] is not None else "  - "
        print(f"{c:<7}{r['dist']:6d}km{r['rr']:6d}km{d:>6}{'是' if r['relevant'] else '否':>5}"
              f"{r['rain']:5d}mm{r['past']:5d}mm{r['soil']:6.2f}{(str(r['gust'])+'km/h') if r['gust'] else '-':>8}{r['level']:3d}"
              f"{r['phase']:>8}{(f"{r['durH']:.0f}h" if r['durH'] else '-'):>7}"
              f"{'⚠ 报' if r['longRain'] else '—':>9}")

    if sweep:
        print("\n=== 时序稳定性推演（同一城市在不同时刻查看）===")
        for c, lat, lng, wx, _ in rows[:3]:
            print(f"\n【{c}】")
            print(f"  {'查看时刻':<18}{'相关':>5}{'过程':>7}{'已下':>7}{'待下':>7}{'档':>3}{'阶段':>8}{'内涝单':>7}")
            for dh in (-60, -48, -36, -24, -12, 0, 12):
                ep = now + dh * 3600
                tab = {"_city": pct_all.get(c)}
                r = assess(lat, lng, pts, fc, wx, ep, tab)
                ts = datetime.fromtimestamp(ep, BJT).strftime("%m-%d %H:%M")
                mark = " ←现在" if dh == 0 else ""
                print(f"  {ts:<18}{'是' if r['relevant'] else '否':>5}{r['rain']:6d}mm"
                      f"{r['past']:6d}mm{r['future']:6d}mm{r['level']:3d}{r['phase']:>9}"
                      f"{'推送' if r['flood'] else '—':>7}{mark}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/sanity_check.mjs
#!/usr/bin/env node
/* 判定口径抽样自检（node 版）——与前端共用 docs/assess-core.js 这**同一个**评估核心。

   为什么换掉 Python 版：sanity_check.py 是 assess() 的平行复刻，同步过五次仍有漂移
   （Pa 公式最近一天权重 panel 用 K¹、Python 用 K⁰）——验收工具与被验收对象是两套
   实现，验收本身不可信。本脚本只负责**取数**（与前端同源：温州台风网 + Open-Meteo），
   判定全部交给 core：core 一改，这里自动跟随，永远不会漂。

   用法（与 Python 版一致）：
     node scripts/sanity_check.mjs 202618
     node scripts/sanity_check.mjs 202618 --sweep
     node scripts/sanity_check.mjs 202610 --at 2026-07-04T12 --cities 南宁市,北海市
     node scripts/sanity_check.mjs 202618 --event --cities 台州市,温州市
                                          # 事件时间线层(Phase B)：整场事件的档位曲线、
                                          # 峰值档与此刻档的关系（「曾4现2」类事实）
*/
import { createRequire } from "node:module";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import path from "node:path";
import vm from "node:vm";

const require = createRequire(import.meta.url);
const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");

// ① 同一把尺子：直接 require 前端的 assess-core.js
const AssessCore = require(path.join(ROOT, "docs", "assess-core.js"));
// ② estGaleRadius 权威表在 data.js：vm 执行拿 TyphoonData（不复制表）
const dataCtx = {};
vm.createContext(dataCtx);
vm.runInContext(readFileSync(path.join(ROOT, "docs", "data.js"), "utf-8") +
  ";globalThis.__T = TyphoonData;", dataCtx);
AssessCore.configure({ estGaleRadius: dataCtx.__T.estGaleRadius });

const BJT_OFF = 8 * 3600e3;
const fmtBJT = (ms) => new Date(ms + BJT_OFF).toISOString().slice(0, 16).replace("T", " ");

async function get(url, tries = 3) {
  for (let i = 0; i < tries; i++) {
    try {
      const r = await fetch(url, { headers: { "User-Agent": "typhoonandcicada-selftest" } });
      const s = await r.text();
      if (s.trim()) return JSON.parse(s);
    } catch (e) { /* retry */ }
    await new Promise((res) => setTimeout(res, 2000));
  }
  throw new Error("接口多次为空: " + url);
}

function loadCityCoords() {
  const reg = JSON.parse(readFileSync(path.join(ROOT, "docs", "data", "regions.json"), "utf-8"));
  const out = {};
  for (const pv of Object.values(reg))
    for (const [city, cv] of Object.entries(pv.cities || {})) out[city] = { lat: cv.lat, lng: cv.lng };
  return out;
}

/* 台风 → data.js normalize 后的形状（track/forecasts），取数逻辑与前端 loadStorm 对齐 */
async function loadStorm(tfid) {
  const d = await get(`https://typhoon.slt.zj.gov.cn/Api/TyphoonInfo/${tfid}`);
  const pts = d.points || [];
  const norm = (p) => ({ time: p.time.replace("T", " ").split("+")[0].split(".")[0],
    lat: +p.lat, lng: +p.lng, strong: p.strong || "", power: p.power || "",
    r7: p.radius7 ? String(p.radius7).split("|").filter(Boolean).map(Number) : null });
  const fcs = {};
  for (let i = pts.length - 1; i >= 0; i--) {
    for (const fc of pts[i].forecast || []) {
      if (fc.tm && !fcs[fc.tm] && (fc.forecastpoints || []).length)
        fcs[fc.tm] = { points: fc.forecastpoints.map(norm) };
    }
    if (Object.keys(fcs).length) break;
  }
  return { name: d.name, storm: { tfid, track: pts.map(norm), forecasts: fcs } };
}

/* 逐时风雨 → 前端 fdata 形状 {t(毫秒),ts,p,g,cur}；daily → soilFromDaily（core 口径） */
function toFdata(hourly) {
  const ts = hourly.time;                                 // "YYYY-MM-DDTHH:MM" 北京钟面
  const t = ts.map((s) => new Date(s + ":00+08:00").getTime());
  return { t, ts, p: hourly.precipitation, g: hourly.wind_gusts_10m, cur: null };
}
async function loadWx(lat, lng) {
  const d = await get(`https://api.open-meteo.com/v1/forecast?latitude=${lat}&longitude=${lng}` +
    `&hourly=precipitation,wind_gusts_10m&daily=precipitation_sum&past_days=14&forecast_days=7&timezone=Asia%2FShanghai`);
  const today = new Date(Date.now() + BJT_OFF).toISOString().slice(0, 10);
  const ante = (d.daily.time || []).flatMap((dd, i) =>
    dd < today && d.daily.precipitation_sum[i] != null ? [d.daily.precipitation_sum[i]] : []);
  return { fdata: toFdata(d.hourly), soil: AssessCore.soilFromDaily(ante) };
}
async function loadWxArchive(lat, lng, atMs) {
  const s = new Date(atMs - 14 * 86400e3 + BJT_OFF).toISOString().slice(0, 10);
  const e = new Date(atMs + 7 * 86400e3 + BJT_OFF).toISOString().slice(0, 10);
  const d = await get(`https://archive-api.open-meteo.com/v1/archive?latitude=${lat}&longitude=${lng}` +
    `&start_date=${s}&end_date=${e}&hourly=precipitation,wind_gusts_10m&daily=precipitation_sum&timezone=Asia%2FShanghai`);
  const cutoff = new Date(atMs + BJT_OFF).toISOString().slice(0, 10);
  const ante = (d.daily.time || []).flatMap((dd, i) =>
    dd < cutoff && d.daily.precipitation_sum[i] != null ? [d.daily.precipitation_sum[i]] : []);
  return { fdata: toFdata(d.hourly), soil: AssessCore.soilFromDaily(ante) };
}

function opt(name) { const i = process.argv.indexOf(name); return i > 0 ? process.argv[i + 1] : null; }
function row(city, a, soil) {
  const pg = a.peakGust ? Math.round(a.peakGust.v) + "km/h" : "-";
  const dur = a.durationH != null ? Math.round(a.durationH) + "h" : "-";
  return [city.padEnd(6), String(Math.round(a.closest.dist)).padStart(5) + "km",
    (a.dirF != null ? "" : ""), a.relevant ? "是" : "否",
    String(a.rain).padStart(4) + "mm", String(a.rainPast ?? 0).padStart(4) + "mm",
    soil.w.toFixed(2), pg.padStart(8), String(a.level), a.phase.padStart(8),
    dur.padStart(5), a.longRain ? "⚠ 报" : "—"].join("  ");
}

const args = process.argv.slice(2).filter((a) => !a.startsWith("--") &&
  a !== opt("--at") && a !== opt("--cities"));
const tfid = args[0] || "202618";
const at = opt("--at");                       // 2026-07-04T12（北京时）
const sweep = process.argv.includes("--sweep");
const DEFAULT_CITIES = ["杭州市", "上海市区", "宁波市", "南京市", "合肥市",
  "福州市", "武汉市", "广州市", "温州市", "青岛市"];
const cities = opt("--cities") ? opt("--cities").split(",").map((s) => s.trim()) : DEFAULT_CITIES;

const coords = loadCityCoords();
let { name, storm } = await loadStorm(tfid);
const nowMs = at ? new Date(`${at}:00:00+08:00`).getTime() : Date.now();

/* 历史回放的忠实构造：track 只到「当时」，之后的实况充当「完美预报」。
   否则 track 是全程历史，slice(-4) 会取到事件结束后的点、预报是事后最后一期，
   fwdClosest / centerAt / strongSoon 的语义全被破坏（首轮验收实测：回放巴威时
   杭州 closest 变成 902km——因为它的最近点在回放时刻的 1 小时后，被错误的
   「末4点」结构漏掉了）。 */
if (at) {
  const s2 = AssessCore.stormAsOf(storm, nowMs);   // 与事件层同一个「当时所知」构造
  if (s2) storm = s2;
}

console.log(`=== ${name} (${tfid}) · ${at ? "历史回放" : "现在"} ${fmtBJT(nowMs)} 北京时 · [node·与前端同核] ===\n`);
console.log(["城市    ", "  距离", "", "相关", " 过程", " 已下", "土湿", "    阵风", "档", "    阶段", " 持续", " 长雨提示"].join("  "));
console.log("-".repeat(86));

const rows = [];
for (const c of cities) {
  if (!coords[c]) { console.log(`  ${c}: 不在区县库，跳过`); continue; }
  const { lat, lng } = coords[c];
  let wx;
  try { wx = at ? await loadWxArchive(lat, lng, nowMs) : await loadWx(lat, lng); }
  catch (e) { console.log(`  ${c}: 天气拉取失败 ${e.message}`); continue; }
  const a = AssessCore.assess({ loc: { lat, lng }, storm, fdata: wx.fdata,
    soilW: wx.soil.w, obs: null, nowT: nowMs });
  rows.push({ c, lat, lng, wx });
  console.log(row(c, a, wx.soil));
}

const evMode = process.argv.includes("--event");
if (evMode) {
  console.log("\n=== 事件时间线层（assessEvent，与前端同核）===");
  const LV = ["", "①关注", "②准备", "③戒备", "④高危"];
  for (const { c, lat, lng, wx } of rows) {
    const ev = AssessCore.assessEvent({ loc: { lat, lng }, storm, fdata: wx.fdata,
      soilW: wx.soil.w, obs: null, nowT: nowMs });
    if (!ev.samples.length) { console.log(`\n【${c}】不相关，无事件`); continue; }
    // 档位曲线压缩显示：连续同档合并为「档×小时数」
    const runs = [];
    for (const sm of ev.samples) {
      const last = runs[runs.length - 1];
      if (last && last.level === sm.level) last.n++;
      else runs.push({ level: sm.level, n: 1 });
    }
    const stepH = Math.round((ev.samples[1].t - ev.samples[0].t) / 3.6e6);
    const curve = runs.map((r) => `${r.level}×${r.n * stepH}h`).join(" → ");
    console.log(`\n【${c}】事件跨度 ${fmtBJT(ev.spanStart)} ~ ${fmtBJT(ev.spanEnd)}（采样 ${ev.samples.length} 点 / ${stepH}h 步长）`);
    console.log(`  档位曲线: ${curve}`);
    console.log(`  事件峰值: ${LV[ev.peakLevel]}（${ev.peakAt ? fmtBJT(ev.peakAt) : "-"}）  此刻: ${LV[ev.base.level]}  12h前: ${LV[ev.levelPrev12h]}` +
      (ev.peakLevel > ev.base.level ? `  ← 已从峰值回落（「曾${ev.peakLevel}现${ev.base.level}」）` : ""));
    // 一致性自检：峰值必须 ≥ 所有采样与此刻档
    const bad = ev.samples.filter((sm) => sm.level > ev.peakLevel).length;
    if (bad || ev.base.level > ev.peakLevel) console.log(`  ⚠️ 一致性失败：存在高于峰值的采样`);
  }
}

if (sweep) {
  console.log("\n=== 时序稳定性推演（同一城市在不同时刻查看）===");
  for (const { c, lat, lng, wx } of rows.slice(0, 3)) {
    console.log(`\n【${c}】`);
    console.log("  查看时刻            相关   过程    已下    待下  档     阶段");
    for (const dh of [-60, -48, -36, -24, -12, 0, 12]) {
      const ep = nowMs + dh * 3600e3;
      const a = AssessCore.assess({ loc: { lat, lng }, storm, fdata: wx.fdata,
        soilW: wx.soil.w, obs: null, nowT: ep });
      const mark = dh === 0 ? " ←现在" : "";
      console.log(`  ${fmtBJT(ep)}   ${a.relevant ? "是" : "否"}  ${String(a.rain).padStart(4)}mm` +
        `  ${String(a.rainPast ?? 0).padStart(4)}mm  ${String(a.rainFuture ?? 0).padStart(4)}mm` +
        `  ${a.level}  ${a.phase.padStart(8)}${mark}`);
    }
  }
}
